imap-proto/src/parser/rfc5464.rs

```diff
@@ -0,0 +1,255 @@
+{
+    if i.len() == l ||
+        i[l] == b' ' ||
+        i[l] == b'\r'
+    {
+        return entry_parse_stage::Done(l);
+    }
+    if i[l] != b'/'
+    {
+        return entry_parse_stage::Fail(nom::Err::Error(nom::Context::Code(i, nom::ErrorKind::IsNot)));
+    }
+    for j in 1..(i.len() - l)
+    {
+        if !is_entry_component_char(i[l+j])
+        {
+            return entry_parse_stage::Path(l+j);
+        }
+    }
+    entry_parse_stage::Done(i.len())
+}
+
+fn check_entry_name(i: &[u8]) -> IResult<&[u8], &[u8]> {
+    let mut stage = entry_parse_stage::Private_Shared(0);
+    loop
+    {
+        match stage
+        {
+            entry_parse_stage::Private_Shared(_) => {println!("PrSh");stage = check_private_shared(i);}
+            entry_parse_stage::Admin(l) => {println!("Adm");stage = check_admin(i, l);}
+            entry_parse_stage::Vendor_Comment(l) => {println!("VeCo");stage = check_vendor_comment(i, l);}
+            entry_parse_stage::Path(l) => {println!("Path");stage = check_path(i, l);}
+            entry_parse_stage::Done(l) => {println!("Entr_name OK");return Ok((&i[l..], &i[..l]));}
+            entry_parse_stage::Fail(e) => {println!("Entr_name FAIL");return std::result::Result::Err(e);}
+        }
+    }
+}
+
+fn entry_name(i: &[u8]) -> IResult<&[u8], &[u8]> {
+    let astring_res = astring(i);
+    match astring_res {
+        Err(e) => { return Err(e); },
+        Ok((i,a)) => {
+            match check_entry_name(a) {
+                Ok(_) => { return Ok((i,a)); },
+                Err(e) => { return Err(e); },
+            }
+        },
+    }
+}
+
+named!(keyval_pair<(&[u8], &[u8])>, do_parse!(
+    key: entry_name >>
+        tag_s!(" ") >>
+        value: quoted >>
+        ((key, value))
+));
+
+named!(keyval_list<Vec<(&[u8], &[u8])>>, do_parse!(
+    list: parenthesized_nonempty_list!(do_parse!(
+        key: entry_name >>
+            tag_s!(" ") >>
+            value: quoted >>
+            (key, value)
+    )) >>
+        (list)
+));
+
+named!(entry_list<Vec<&[u8]>>, do_parse!(
+    list: separated_list!(tag_s!(" "), entry_name) >>
+        (list)
+));
+
+named!(metadata_common<&[u8]>, do_parse!(
+    tag_s!("METADATA ") >>
+        mbox: quoted >>
+        tag_s!(" ") >>
+        (mbox)
+));
+
+named!(pub metadata_solicited<Vec<(&[u8],&[u8])>>, do_parse!(
+    mbox: metadata_common >>
+    tail: keyval_list >>
+        (tail)
+));
+
+named!(pub metadata_unsolicited<Vec<&[u8]>>, do_parse!(
+    mbox: metadata_common >>
+        tail: entry_list >>
+        (tail)
+));
+
+#[cfg(test)]
+mod tests {
+    use nom;
+    use core::*;
+    use super::{metadata_solicited, metadata_unsolicited};
+    use super::entry_name;
+    use types::*;
+
+    #[test]
+    fn test_solicited_fail_1() {
+        match metadata_solicited(b"METADATA \"\" (/asdfg \"asdf\")\r\n") {
+            Err(_) => {},
+            _ => panic!("Error required when entry name is not starting with /private or /shared"),
```

Comment:
- use `assert!(metadata_solicited(...).is_err())` instead of panicking.
- you can use part of the panic message as test function name to describe what you are testing.